File: server/src/api/playback.ts
import type { Request, Response } from 'express'
import { promisify } from 'util'
import { gzip } from 'zlib'
import { prisma } from '../utils/db'

const gzipAsync = promisify(gzip)

interface AuthUser {
    id: string
    canReadAllRooms: boolean
    canWriteAllRooms: boolean
    canDeleteAllRooms: boolean
}

function hasGlobalRead(user: AuthUser): boolean {
    return user.canReadAllRooms || user.canWriteAllRooms || user.canDeleteAllRooms
}

export async function getPlaybackUpdates(req: Request, res: Response) {
    try {
        const authUser = (req as any).authUser as AuthUser | undefined
        if (!authUser) {
            return res.status(401).json({ error: 'Authentication required' })
        }

        const { roomId } = req.params

        // Get room
        const room = await prisma.room.findUnique({
            where: { id: roomId },
            include: {
                participants: {
                    select: { userId: true },
                },
            },
        })

        if (!room) {
            return res.status(404).json({ error: 'Room not found' })
        }

        if (!room.isEnded) {
            return res.status(400).json({ error: 'Room has not ended yet' })
        }

        const isOwner = room.ownerId === authUser.id
        const isParticipant = room.participants.some(p => p.userId === authUser.id)

        if (!hasGlobalRead(authUser) && !isOwner && !isParticipant) {
            return res.status(403).json({ error: 'Access denied' })
        }

        // Get all updates for this document
        const updates = await prisma.documentUpdate.findMany({
            where: {
                documentId: room.documentId,
            },
            orderBy: {
                timestamp: 'asc',
            },
            select: {
                id: true,
                timestamp: true,
                update: true,
                userId: true,
            },
        })

        if (updates.length === 0) {
            return res.json({
                updates: [],
                startTime: null,
                endTime: null,
                duration: 0,
            })
        }

        const startTime = updates[0]!.timestamp
        const endTime = updates[updates.length - 1]!.timestamp
        const duration = (endTime.getTime() - startTime.getTime()) / 1000 // seconds

        // Compress and convert updates to base64 for transport
        const updatesData = await Promise.all(
            updates.map(async (u) => {
                const compressed = await gzipAsync(u.update)
                return {
                    id: u.id,
                    timestamp: u.timestamp.toISOString(),
                    update: compressed.toString('base64'),
                    userId: u.userId,
                }
            })
        )

        res.json({
            updates: updatesData,
            startTime: startTime.toISOString(),
            endTime: endTime.toISOString(),
            duration,
        })
    } catch (error) {
        console.error('Get playback updates error:', error)
        res.status(500).json({ error: 'Internal server error' })
    }
}

File: server-rs/src/routes/playback.rs
use axum::extract::{Path, State};
use axum::Json;
use base64::{engine::general_purpose::STANDARD as BASE64_STANDARD, Engine as _};
use flate2::{write::GzEncoder, Compression};
use serde_json::json;
use std::io::Write;

use crate::{
    auth::AuthUser,
    db::db_error,
    error::ApiError,
    models::DocumentUpdateRow,
    permissions::has_global_read,
    state::AppState,
    utils::time::{to_iso_string, to_iso_string_opt},
};

#[derive(Debug, sqlx::FromRow)]
struct RoomAccessRow {
    id: String,
    owner_id: String,
    is_ended: bool,
}

pub async fn get_playback_updates(
    State(state): State<AppState>,
    auth_user: AuthUser,
    Path(room_id): Path<String>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let room = sqlx::query_as::<_, RoomAccessRow>(
        r#"
        SELECT id, "ownerId" as owner_id, "isEnded" as is_ended
        FROM "Room"
        WHERE id = $1
        "#
    )
    .bind(&room_id)
    .fetch_optional(&state.db)
    .await
    .map_err(|err| db_error(err, "Failed to load room"))?;

    let room = match room {
        Some(room) => room,
        None => return Err(ApiError::not_found("Room not found")),
    };

    let is_owner = room.owner_id == auth_user.id;
    let is_privileged = auth_user.role == "admin"
        || auth_user.role == "superuser"
        || has_global_read(&auth_user);

    if !is_owner && !is_privileged {
        return Err(ApiError::not_found("Room not found"));
    }

    if !room.is_ended {
        return Err(ApiError::bad_request("Room has not ended yet"));
    }

    let updates = sqlx::query_as::<_, DocumentUpdateRow>(
        r#"
        SELECT
            id,
            "documentId" as document_id,
            update,
            timestamp,
            "userId" as user_id
        FROM "DocumentUpdate"
        WHERE "documentId" = $1
        ORDER BY timestamp ASC
        "#,
    )
    .bind(&room.id)
    .fetch_all(&state.db)
    .await
    .map_err(|err| db_error(err, "Failed to load playback updates"))?;

    if updates.is_empty() {
        return Ok(Json(json!({
            "updates": [],
            "startTime": null,
            "endTime": null,
            "duration": 0,
        })));
    }

    let start_time = updates.first().map(|u| u.timestamp).unwrap();
    let end_time = updates.last().map(|u| u.timestamp).unwrap();
    let duration = (end_time - start_time).num_milliseconds() as f64 / 1000.0;

    let mut compressed_updates = Vec::with_capacity(updates.len());
    for update in updates {
        let compressed = gzip_bytes(&update.update)
            .map_err(|err| ApiError::internal(format!("Failed to compress update: {err}")))?;
        let encoded = BASE64_STANDARD.encode(compressed);
        compressed_updates.push(json!({
            "id": update.id,
            "timestamp": to_iso_string(update.timestamp),
            "update": encoded,
            "userId": update.user_id,
        }));
    }

    Ok(Json(json!({
        "updates": compressed_updates,
        "startTime": to_iso_string_opt(Some(start_time)),
        "endTime": to_iso_string_opt(Some(end_time)),
        "duration": duration,
    })))
}

fn gzip_bytes(input: &[u8]) -> Result<Vec<u8>, std::io::Error> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(input)?;
    encoder.finish()
}
